File: client/src/core/sync-engine.ts
import { createHash } from "node:crypto";
import { nanoid } from "nanoid";
import type { ClipboardPayload } from "../types.js";
import { isAllowedByPolicy } from "./policy-engine.js";
import type { SharePolicy } from "../types.js";

export interface SyncDecision {
  accepted: boolean;
  reason?: string;
}

export class SyncEngine {
  private readonly localDeviceId: string;

  private readonly recentlySeen = new Set<string>();

  constructor(localDeviceId: string) {
    this.localDeviceId = localDeviceId;
  }

  makeItemId(content: Uint8Array, createdAtUnix: number): string {
    const digest = createHash("sha256").update(content).update(String(createdAtUnix)).digest("hex");
    return `item_${digest.slice(0, 16)}_${nanoid(6)}`;
  }

  shouldSend(payload: ClipboardPayload, policy: SharePolicy): SyncDecision {
    if (!isAllowedByPolicy(policy, payload)) {
      return { accepted: false, reason: "blocked_by_policy" };
    }

    if (this.recentlySeen.has(payload.itemId)) {
      return { accepted: false, reason: "duplicate_item" };
    }

    this.markSeen(payload.itemId);
    return { accepted: true };
  }

  shouldApplyIncoming(payload: ClipboardPayload): SyncDecision {
    if (payload.sourceDeviceId === this.localDeviceId) {
      return { accepted: false, reason: "loopback" };
    }

    if (this.recentlySeen.has(payload.itemId)) {
      return { accepted: false, reason: "duplicate_item" };
    }

    this.markSeen(payload.itemId);
    return { accepted: true };
  }

  private markSeen(itemId: string): void {
    this.recentlySeen.add(itemId);
    if (this.recentlySeen.size > 1000) {
      const [first] = this.recentlySeen;
      if (first) {
        this.recentlySeen.delete(first);
      }
    }
  }
}

File: apps/mobile-android/app/src/main/kotlin/dev/sharepaste/android/data/SyncEngine.kt
package dev.sharepaste.android.data

import dev.sharepaste.android.data.model.ClipboardPayload
import dev.sharepaste.android.data.model.PayloadKind
import dev.sharepaste.android.data.model.SharePolicyUi
import java.security.MessageDigest
import java.util.UUID

class SyncEngine(private val localDeviceId: () -> String) {
    private val recentlySeen = LinkedHashSet<String>()

    fun makeItemId(content: ByteArray, createdAtUnix: Long): String {
        val digest = MessageDigest.getInstance("SHA-256")
            .digest(content + createdAtUnix.toString().toByteArray())
            .joinToString("") { byte -> "%02x".format(byte) }
        return "item_${digest.take(16)}_${UUID.randomUUID().toString().take(6)}"
    }

    fun shouldSend(payload: ClipboardPayload, policy: SharePolicyUi): Boolean {
        if (!isAllowedByPolicy(policy, payload.kind, payload.sizeBytes)) {
            return false
        }
        if (recentlySeen.contains(payload.itemId)) {
            return false
        }
        markSeen(payload.itemId)
        return true
    }

    fun shouldApplyIncoming(payload: ClipboardPayload): Boolean {
        if (payload.sourceDeviceId == localDeviceId()) {
            return false
        }
        if (recentlySeen.contains(payload.itemId)) {
            return false
        }
        markSeen(payload.itemId)
        return true
    }

    private fun isAllowedByPolicy(policy: SharePolicyUi, kind: PayloadKind, sizeBytes: Long): Boolean {
        return when (kind) {
            PayloadKind.TEXT -> policy.allowText
            PayloadKind.IMAGE -> policy.allowImage
            PayloadKind.FILE -> policy.allowFile && sizeBytes < policy.maxFileSizeBytes
        }
    }

    private fun markSeen(itemId: String) {
        recentlySeen += itemId
        if (recentlySeen.size > 1000) {
            val first = recentlySeen.firstOrNull() ?: return
            recentlySeen.remove(first)
        }
    }
}
